sqlstats: fix data race between collector and ingester

dhartunian opened the pull request:
I've admittedly not been able to reproduce the data race observed in the same `*sqlstats.RecordedStmtStats` object in two different locations. One of these is now redunant. The transactionID set in the `StatsCollector` happens before the data is processed by the `SQLStatsIngester` so the setting of the transactionID in the ingester is removed. If it happens to get statements without corresponding transactionIDs, that should be a bug.

The reason this was introduced was likely
https://github.com/cockroachdb/cockroach/pull/141767 which unified the types used by the two components. Previously, they were different structs and ownership was clear and required editing the transactionID twice.

This PR also introduces a `doc.go` file to the `sslocal` package in an effort to guide the reviewer and provide context for future work. The diagram reflects the changes in this commit.

Resolves: #146796

Release note: None

dhartunian committed:
sqlstats: fix data race between collector and ingester

I've admittedly not been able to reproduce the data race observed in
the same `*sqlstats.RecordedStmtStats` object in two different
locations. One of these is now redunant. The transactionID set in
the `StatsCollector` happens before the data is processed by the
`SQLStatsIngester` so the setting of the transactionID in the ingester
is removed. If it happens to get statements without corresponding
transactionIDs, that should be a bug.

The reason this was introduced was likely
https://github.com/cockroachdb/cockroach/pull/141767 which unified
the types used by the two components. Previously, they were different
structs and ownership was clear and required editing the transactionID
twice.

This PR also introduces a `doc.go` file to the `sslocal` package in an
effort to guide the reviewer and provide context for future work. The
diagram reflects the changes in this commit.

Resolves: #146796

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147133)
<!-- Reviewable:end -->

dhartunian committed:
sqlstats: add test for collector -> ingester flow

This test exercises the recording of statements and transactions
with the `StatsCollector`. We record a bunch of statements in a
transaction, then close that transaction and make sure that the
ingester saw all 100 statements and that they were tagged correctly
with the transaction's fingerprint ID.

Release note: None

dhartunian commented:
TFTR!

bors r=alyshanjahani-crl
